geomstats/geometry/heisenberg.py

```diff
@@ -0,0 +1,204 @@
+                                  point_b[0]]))
+
+    def inverse(self, point):
+        """Compute the group inverse of point.
+
+        Parameters
+        ----------
+        point : array-like, shape=[..., 3]
+            Point.
+
+        Returns
+        -------
+        inv_point : array-like, shape=[..., 3]
+            Inverse.
+        """
+        return -point
+
+    def jacobian_translation(
+            self, point, left_or_right='left'):
+        """Compute the Jacobian of left/right translation by a point.
+
+        Compute the Jacobian matrix of the left translation by the point.
+
+        Parameters
+        ----------
+        point : array-like, shape=[..., {dim, [n, n]]
+            Point.
+        left_or_right : str, {'left', 'right'}
+            Indicate whether to calculate the differential of left or right
+            translations.
+            Optional, default: 'left'.
+
+        Returns
+        -------
+        jacobian : array-like, shape=[..., dim, dim]
+            Jacobian of the left/right translation by point.
+        """
+        if left_or_right == 'left':
+            return(gs.array([[1, 0, 0],
+                             [0, 1, 0],
+                             [-point[1] / 2, point[0] / 2, 1]]))
+        else:
+            return(gs.array([[1, 0, 0],
+                             [0, 1, 0],
+                             [point[1] / 2, -point[0] / 2, 1]]))
+
+    def random_point(self, n_samples=1, bound=1.):
+        """Sample in the Euclidean space R^3 with a uniform distribution in a box.
+
+        Parameters
+        ----------
+        n_samples : int
+            Number of samples.
+            Optional, default: 1.
+        bound : float
+            Side of hypercube support of the uniform distribution.
+            Optional, default: 1.0
+
+        Returns
+        -------
+        point : array-like, shape=[..., dim]
+           Sample.
+        """
+        size = (self.dim,)
+        if n_samples != 1:
+            size = (n_samples, self.dim)
+        point = bound * (gs.random.rand(*size) - 0.5) * 2
+
+        return point
+
+    def exp_from_identity(self, tangent_vec):
+        """Compute the group exponential of the tangent vector at the identity.
+
+        Parameters
+        ----------
+        tangent_vec : array-like, shape=[..., dimension]
+            Tangent vector at base point.
+
+        Returns
+        -------
+        point : array-like, shape=[..., dimension]
+            Point.
+        """
+        return tangent_vec
+
+    def log_from_identity(self, point):
+        """Compute the group logarithm of the point at the identity.
+
+        Parameters
+        ----------
+        point : array-like, shape=[..., dimension]
+            Point.
+
+        Returns
+        -------
+        tangent_vec : array-like, shape=[..., dimension]
+            Group logarithm.
+        """
+        return point
+
+    def upperTriangular_matrix_from_vector(self, point):
```

Comment:
I think this method is already implemented in the backend, or will be added by #1142

Reply:
Should I then keep this method ('upperTriangular_matrix_from_vector') in the Heisenberg class, but use the implementation from the backend or #1142, or should the method not be part of the Heisenberg class at all..?

Many thanks for you quick feedback! I'm doing the corrections now

Reply:
Just to be sure:   upperTriangular_matrix_from_vector   maps the vector 'point' to the matrix 

[[1, point[0], point[2] + 1 / 2 * point[0] * point[1]],
 [0, 1, point[1]],
 [0, 0, 1]] 

Are you sure that is already implemented in backend/#1142 ?

Reply:
Ah ok, no there is the method `SymmetricMatrices.from_vector` that maps `[1, a, b, 1, c, 1]` to matrix `[[1, a, b], [a, 1, c], [b, c, 1]]` and that is vectorized to work with multiple outputs, and then there is `triu` to keep only the upper part.

Reply:
Ok thanks I'll look at that, and vectorize it